lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -0,0 +1,232 @@
+    OpenSearchPluginInstallError,
+    OpenSearchPluginMissingDepsError,
+    OpenSearchPluginRemoveError,
+    PluginState,
+)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "da838485175f47dbbbb83d76c07cab4c"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+ConfigExposedPlugins = {}
+
+
+class OpenSearchPluginManager:
+    """Manages plugins."""
+
+    def __init__(self, charm, plugins_path: str = None):
+        """Creates the plugin manager object based on the charm and home_path.
+
+        Stores the home path and, optionally, plugins path can also be passed if it is
+        not available in {home_path}/plugins.
+        """
+        self._charm = charm
+        self._opensearch = charm.opensearch
+        self._opensearch_config = charm.opensearch_config
+        self._charm_config = self._charm.model.config
+        self._plugins_path = plugins_path
+        self._keystore = charm.opensearch_keystore
+
+    @property
+    def plugins(self) -> List[OpenSearchPlugin]:
+        """Returns List of installed plugins."""
+        return [
+            plugin_data["class"](self._plugins_path, extra_config=self._extra_conf(plugin_data))
+            for _, plugin_data in ConfigExposedPlugins.items()
+        ]
+
+    def _extra_conf(self, plugin_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
+        """Returns either the config or relation data for the target plugin."""
+        relation_name = plugin_data.get("relation")
+        relation = self._charm.model.get_relation(relation_name) if relation_name else None
+        if not relation:
+            return None
+        return relation.data[self._charm.app]
+
+    def run(self) -> bool:
+        """Runs a check on each plugin: install, execute config changes or remove.
+
+        This method should be called at config-changed event. Returns if needed restart.
+        """
+        restart_needed = False
+        for plugin in self.plugins:
+            restart_needed = any(
+                [
+                    self._install_if_needed(plugin),
+                    self._configure_if_needed(plugin),
+                    self._disable_if_needed(plugin),
+                    restart_needed,
+                ]
+            )
+        return restart_needed
+
+    def _install_if_needed(self, plugin: OpenSearchPlugin) -> bool:
+        """Installs all the plugins enabled via the config/relation.
+
+        Check if plugin in status: PluginState.MISSING and config/relation is set.
+        Returns True if a restart is needed.
+        """
+        # installed is needed because we may have a plugin already installed
+        # i.e. we fail the install command but we do not need to do anything else.
+        installed = False
```

Comment:
nits:
looks like we can simply return "True"  in line 123 and False in line 124